% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/MatrixGeneScores.R
\name{addGeneScoreMatrix}
\alias{addGeneScoreMatrix}
\title{Add GeneScoreMatrix to ArrowFiles or an ArchRProject}
\usage{
addGeneScoreMatrix(input = NULL, genes = getGenes(input),
  geneModel = "exp(-abs(x)/5000) + exp(-1)",
  matrixName = "GeneScoreMatrix", extendUpstream = c(1000, 1e+05),
  extendDownstream = c(1000, 1e+05), geneUpstream = 5000,
  geneDownstream = 0, useGeneBoundaries = TRUE, useTSS = FALSE,
  tileSize = 500, ceiling = 4, geneScaleFactor = 5,
  scaleTo = 10000, excludeChr = c("chrY", "chrM"),
  blacklist = getBlacklist(input), threads = getArchRThreads(),
  parallelParam = NULL, subThreading = TRUE, force = FALSE)
}
\arguments{
\item{input}{An \code{ArchRProject} object or character vector of ArrowFiles.}

\item{genes}{A stranded \code{GRanges} object containing the ranges associated with all gene start and end coordinates.}

\item{geneModel}{A string giving a "gene model function" used for weighting peaks for gene score calculation. This string
should be a function of \code{x}, where \code{x} is the stranded distance from the transcription start site of the gene.}

\item{matrixName}{The name to be used for storage of the gene activity score matrix in the provided \code{ArchRProject} or ArrowFiles.}

\item{extendUpstream}{The minimum and maximum number of basepairs upstream of the transcription start site to consider for gene
activity score calculation.}

\item{extendDownstream}{The minimum and maximum number of basepairs downstream of the transcription start site to consider for gene activity score calculation.}

\item{useGeneBoundaries}{A boolean value indicating whether gene boundaries should be employed during gene activity score
calculation. Gene boundaries refers to the process of preventing tiles from contributing to the gene score of a given gene
if there is a second gene's transcription start site between the tile and the gene of interest.}

\item{tileSize}{The size of the tiles used for binning counts prior to gene activity score calculation.}

\item{ceiling}{The maximum counts per tile allowed. This is used to prevent large biases in tile counts.}

\item{scaleTo}{Each column in the calculated gene score matrix will be normalized to a column sum designated by \code{scaleTo}.}

\item{excludeChr}{A character vector containing the \code{seqnames} of the chromosomes that should be excluded from this analysis.}

\item{blacklist}{A \code{GRanges} object containing genomic regions to blacklist that may be extremeley over-represented and thus
biasing the geneScores for genes nearby that locus.}

\item{threads}{The number of threads to be used for parallel computing.}

\item{parallelParam}{A list of parameters to be passed for biocparallel/batchtools parallel computing.}

\item{subThreading}{A boolean determining whether possible use threads within each multi-threaded subprocess if greater than the number of input samples.}

\item{force}{A boolean value indicating whether to force the matrix indicated by \code{matrixName} to be overwritten if it already exist in the given \code{input}.}
}
\description{
This function, for each sample, will independently compute counts for each tile
per cell and then infer gene activity scores.
}
