% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AnnotationPeaks.R
\name{addMotifAnnotations}
\alias{addMotifAnnotations}
\title{Add motif annotations to an ArchRProject}
\usage{
addMotifAnnotations(
  ArchRProj = NULL,
  motifSet = "cisbp",
  annoName = "Motif",
  species = NULL,
  collection = "CORE",
  motifPWMs = NULL,
  cutOff = 5e-05,
  width = 7,
  version = 2,
  force = FALSE,
  logFile = createLogFile("addMotifAnnotations"),
  ...
)
}
\arguments{
\item{ArchRProj}{An \code{ArchRProject} object.}

\item{motifSet}{The name of a curated motif set to be used for annotation. Options include: (i) "JASPAR2016", "JASPAR2018", "JASPAR2020"
which gives the 2016, 2018, or 2020 version of JASPAR motifs, (ii) one of "cisbp", "encode", or "homer" which gives the
corresponding motif sets from the \code{chromVAR} package, or (iii) "vierstra" which gives the clustered archetype motifs
created by Jeff Vierstra (https://github.com/jvierstra/motif-clustering).}

\item{annoName}{The name of the \code{peakAnnotation} object to be stored in the provided \code{ArchRProject}}

\item{species}{The latin name of the species relevant to the supplied \code{ArchRProject}. This is used for identifying which motif to be
used from CisBP/JASPAR. For JASPAR, \code{species} is passed to \code{TFBS::getMatrixSet} and some species names are not recognized. In these cases
it is possible to use the NCBI taxonomy ID. By default, this function will attempt to guess the species based on the value from \code{getGenome()}.}

\item{collection}{If one of the JASPAR motif sets is used via \code{motifSet}, this parameter allows you to indicate the JASPAR
collection to be used. See \code{getMatrixSet()} from \code{TFBSTools} for all options to supply for collection. If \code{motifSet} is
"vierstra", then this must either be "archetype" (for the v2.1 clustered models) or "individual" (for the original v1 individual motif models).
NOTE: vierstra archetype motifs are currently in beta and have not been finalized by Jeff Vierstra.}

\item{motifPWMs}{A custom set of motif PWMs as a PWMatrixList to be used instead of \code{motifSet} for adding motif annotations.
If \code{motifPWMs} is used, \code{motifSet} will be ignored.}

\item{cutOff}{The p-value cutoff to be used for motif search. The p-value is determined vs a background set of sequences
(see \code{MOODS} for more details on this determination).}

\item{width}{The width in basepairs to consider for motif matches. See the \code{motimatchr} package for more information.}

\item{version}{An integer specifying version 1 or version 2 of chromVARmotifs see github for more info GreenleafLab/chromVARmotifs.}

\item{force}{A boolean value indicating whether to force the \code{peakAnnotation} object indicated by \code{annoName} to be overwritten if
it already exists in the given \code{ArchRProject}.}

\item{logFile}{The path to a file to be used for logging ArchR output.}

\item{...}{Additional parameters to be passed to \code{TFBSTools::getMatrixSet} for getting a JASPAR PWM object.}
}
\description{
This function adds information about which peaks contain motifs to a given ArchRProject. For each peak, a binary value
is stored indicating whether each motif is observed within the peak region.
}
\examples{

# Get Test ArchR Project
proj <- getTestProject()

# Add Motif Annotations
proj <- addMotifAnnotations(proj, motifSet = "cisbptest", annoName = "test")

}
