% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ReproduciblePeakSet.R
\name{addReproduciblePeakSet}
\alias{addReproduciblePeakSet}
\title{Add a Reproducible Peak Set to an ArchRProject}
\usage{
addReproduciblePeakSet(ArchRProj = NULL, groupBy = "Clusters",
  reproducibility = "2", peaksPerCell = 500, maxPeaks = 150000,
  minCells = 25, excludeChr = c("chrM", "chrY"),
  pathToMacs2 = findMacs2(), genomeSize = NULL, shift = -75,
  extsize = 150, method = "q", cutOff = 0.1, extendSummits = 250,
  promoterDist = 500, genomeAnno = getGenomeAnnotation(ArchRProj),
  geneAnno = getGeneAnnotation(ArchRProj),
  additionalParams = "--nomodel --nolambda", threads = 1,
  parallelParam = "mclapply", force = FALSE, verboseHeader = TRUE,
  verboseAll = FALSE, ...)
}
\arguments{
\item{ArchRProj}{An \code{ArchRProject} object.}

\item{groupBy}{The name of the column in \code{cellColData} to use for grouping cells together for peak calling.}

\item{reproducibility}{A string that indicates how peak reproducibility should be handled. This string is dynamic and can be a function of \code{n} where \code{n} is the number of samples being assessed. For example, \code{reproducibility = "2"} means at least 2 samples must have a peak call at this locus and \code{reproducibility = "(n+1)/2"} means that the majority of samples must have a peak call at this locus.}

\item{peaksPerCell}{The limit of number of peaks that can be identified per cell (this is useful for controlling how many peaks can be called from low cell groups).}

\item{maxPeaks}{A numeric threshold for the maximum peaks to retain per group in \code{groupBy} in the union reproducible peak set.}

\item{minCells}{The minimum number of unique cells that was used to create the coverage files on which peaks are called. This is important to allow for exclusion of pseudo-bulk replicates derived from very low cell numbers.}

\item{excludeChr}{A character vector containing the \code{seqnames} of the chromosomes that should be excluded from peak calling.}

\item{pathToMacs2}{The full path to the MACS2 executable.}

\item{genomeSize}{The genome size to be used for MACS2 peak calling (see MACS2 documentation).}

\item{shift}{The number of basepairs to shift each Tn5 insertion. When combined with \code{extsize} this allows you to create proper fragments, centered at the Tn5 insertion site, for use with MACS2 (see MACS2 documentation).}

\item{extsize}{The number of basepairs to extend the MACS2 fragment after \code{shift} has been applied. When combined with \code{extsize} this allows you to create proper fragments, centered at the Tn5 insertion site, for use with MACS2 (see MACS2 documentation).}

\item{method}{The method to use for significance testing in MACS2. Options are "p" for p-value and "q" for q-value. When combined with \code{cutOff} this gives the method and significance threshold for peak calling (see MACS2 documentation).}

\item{cutOff}{The numeric significance cutoff for the testing method indicated by \code{method} (see MACS2 documentation).}

\item{extendSummits}{The number of basepairs to extend peak summits (in both directions) to obtain final fixed-width peaks. For example, \code{extendSummits = 250} will create 501-bp fixed-width peaks from the 1-bp summits.}

\item{promoterDist}{The maximum distance in basepairs from the peak summit to the nearest transcriptional start site to allow for a peak to be annotated as a "promoter" peak.}

\item{genomeAnno}{The genomeAnnotation (see createGenomeAnnotation) is used for downstream analyses for genome information such as nucleotide information (GC info) or chromosome sizes.}

\item{geneAnno}{The geneAnnotation (see createGeneAnnotation) is used for peak labeling as promoter etc.}

\item{additionalParams}{A string of additional parameters to pass to MACS2 (see MACS2 documentation).}

\item{threads}{The number of threads to be used for parallel computing.}

\item{parallelParam}{A list of parameters to be passed for biocparallel/batchtools parallel computing.}

\item{force}{A boolean value indicating whether to force the reproducible peak set to be overwritten if it already exist in the given \code{ArchRProject} peakSet.}

\item{verboseHeader}{A boolean value that determines whether standard output includes verbose sections.}

\item{verboseAll}{A boolean value that determines whether standard output includes verbose subsections.}

\item{...}{additional args}
}
\description{
This function will get insertions from coverage files, call peaks,
and merge peaks to get a "Union Reproducible Peak Set"
}
