<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Getting Started With ArchR • ArchR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Getting Started With ArchR">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ArchR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/ArchR.html">Tutorial</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Articles/doubletRemoval.html">Doublet Removal</a>
    </li>
    <li>
      <a href="../articles/Articles/qcFilters.html">QC Filters</a>
    </li>
    <li>
      <a href="../articles/Articles/peakCalling.html">Peak Calling</a>
    </li>
    <li>
      <a href="../articles/Articles/dimReduction.html">Dimensionality Reduction</a>
    </li>
    <li>
      <a href="../articles/Articles/geneScores.html">Gene Scores and Marker Genes</a>
    </li>
    <li>
      <a href="../articles/Articles/signalTracks.html">Signal Track Ploting</a>
    </li>
    <li>
      <a href="../articles/Articles/footprinting.html">TF Footprinting</a>
    </li>
    <li>
      <a href="../articles/Articles/pseudobulkGeneration.html">Pseudobulk Generation</a>
    </li>
    <li>
      <a href="../articles/Articles/annotations.html">Gene and Genome Annotation</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/faq.html">FAQ</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/GreenleafLab/ArchR">
    <span class="fa fa-github"></span>
     
    ArchR on GitHub
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Getting Started With ArchR</h1>
            
      
      
      <div class="hidden name"><code>ArchR.Rmd</code></div>

    </div>

    
    
<p>The following tutorial shows the basics of setting up and interacting with an ArchR Project using a gold-standard dataset of hematopoietic cells ( CITATION ). This tutorial and all of the accompanying vignettes assume that you are running ArchR <strong>locally</strong>. Once all of these steps work for you, feel free to set up ArchR to work in a cluster environment. This tutorial does not explain every detail of every step. Please see the <a href="articles/index.html">Vignettes section</a> for more details on each major analytical step and all of the major features of ArchR.</p>
<div id="what-is-an-arrowfile-archrproject" class="section level3">
<h3 class="hasAnchor">
<a href="#what-is-an-arrowfile-archrproject" class="anchor"></a>What is an ArrowFile / ArchRProject?</h3>
<p>The base unit of an analytical project in ArchR is called an <code>ArrowFile</code>. Each <code>ArrowFile</code>, stores all of the data associated with an individual sample. Here, a sample would be the most detailed unit of analysis desired (for ex. a single replicate of a particular condition). During creation and as additional analyses are performed, ArchR updates and edits each <code>ArrowFile</code> to contain additional layers of information. Then, an <code>ArchRProject</code> allows you to associate these <code>ArrowFiles</code> together into a single analytical framework.</p>
<div class="figure">
<img src="ArchRProject_Schematic.png" width="700">
</div>
<p>Certain actions can be taken directly on <code>ArrowFiles</code> while other actions are taken on an <code>ArchRProject</code> which in turn updates each associated <code>ArrowFile</code>.</p>
<p>[INSERT ARROW / ARCHRPROJECT DIAGRAM HERE]</p>
</div>
<div id="getting-set-up" class="section level3">
<h3 class="hasAnchor">
<a href="#getting-set-up" class="anchor"></a>Getting Set Up</h3>
<p>The first thing we do is set up our working directory, load our genome annotations, and set the number of threads we would like to use. Depending on the configuration of your local environment, you may need to modify the number of <code>threads</code> used below.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Input Libraries</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ArchR)

<span class="co">#Set/Create Working Directory to Folder</span>
<span class="kw"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span>(<span class="st">""</span>,<span class="dt">recursive =</span> <span class="ot">TRUE</span>)
<span class="kw"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span>(<span class="st">""</span>)

<span class="co">#Load Genome Annotations</span>
<span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"geneAnnoHg19"</span>)
<span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"genomeAnnoHg19"</span>)
geneAnno &lt;-<span class="st"> </span>geneAnnoHg19
genomeAnno &lt;-<span class="st"> </span>genomeAnnoHg19

<span class="co">#Set Threads to be used</span>
threads &lt;-<span class="st"> </span><span class="dv">8</span></code></pre></div>
</div>
<div id="creating-arrow-files" class="section level3">
<h3 class="hasAnchor">
<a href="#creating-arrow-files" class="anchor"></a>Creating Arrow Files</h3>
<p>For this tutorial, we will download a collection of fragment files. Fragment files are one of the base file types of the 10x Genomics analytical platform and can be easily created from any bam file. See QQQ for information on making your own fragment files. Once we have our fragment files, we provide their names as a vector to <code>createArrowFiles</code>. During creation, some basic matrices and data is added to each <code>ArrowFile</code> including a <code>TileMatrix</code> containing insertion counts across genome-wide 500-bp bins.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Get Tutorial Data ~2.2GB To Download</span>
inputFiles &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getTutorialData.html">getTutorialData</a></span>(<span class="st">"Hematopoiesis"</span>)

<span class="co">#Create Arrow Files (~10-15 minutes)</span>
ArrowFiles &lt;-<span class="st"> </span><span class="kw"><a href="../reference/createArrowFiles.html">createArrowFiles</a></span>(
  <span class="dt">inputFiles =</span> inputFiles,
  <span class="dt">sampleNames =</span> <span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(inputFiles),
  <span class="dt">geneAnno =</span> geneAnno,
  <span class="dt">genomeAnno =</span> genomeAnno,
  <span class="dt">threads =</span> threads,
  <span class="dt">force =</span> <span class="ot">FALSE</span>
)</code></pre></div>
</div>
<div id="tidying-up-our-data-and-creating-an-archrproject" class="section level3">
<h3 class="hasAnchor">
<a href="#tidying-up-our-data-and-creating-an-archrproject" class="anchor"></a>Tidying up our data and creating an <code>ArchRProject</code>
</h3>
<p>One major source of trouble in single-cell data is the contribution of “doublets” to the analysis. A doublet refers to a single droplet that received a single barcoded bead and more than one nucleus. This causes the reads from more than one cell to appear as a single cell. We remove these computationally and describe this doublet removal process in more depth in the <a href="articles/Articles/doubletRemoval.html">doublet removal vignette</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Add Infered Doublet Scores to each Arrow File (~5-10 minutes)</span>
doubScores &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addDoubletScores.html">addDoubletScores</a></span>(ArrowFiles, <span class="dt">threads =</span> threads)

<span class="co">#Create ArchRProject</span>
proj &lt;-<span class="st"> </span><span class="kw">ArchRProject</span>(
  <span class="dt">ArrowFiles =</span> ArrowFiles, 
  <span class="dt">geneAnnotation =</span> geneAnno,
  <span class="dt">genomeAnnotation =</span> genomeAnno,
  <span class="dt">outputDirectory =</span> <span class="st">"Heme_Tutorial"</span>
)

<span class="co">#Filter Doublets</span>
proj &lt;-<span class="st"> </span><span class="kw"><a href="../reference/filterDoublets.html">filterDoublets</a></span>(proj)</code></pre></div>
</div>
<div id="dimensionality-reduction" class="section level3">
<h3 class="hasAnchor">
<a href="#dimensionality-reduction" class="anchor"></a>Dimensionality Reduction</h3>
<p>At this point, we have an ArchR project that is ready to be used in downstream visualizations and analyses. The first thing we will do is use an iterative latent semantic indexing (LSI) approach to define clusters in our data. Once we have identified clusters in our data, we can plot a UMAP embedding. For more details, see the <a href="articles/Articles/dimReduction.html">dimensionality reduction vignette</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Reduce Dimensions with Iterative LSI (~5-10 minutes)</span>
proj &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addIterativeLSI.html">addIterativeLSI</a></span>(
  <span class="dt">ArchRProj =</span> proj, 
  <span class="dt">useMatrix =</span> <span class="st">"TileMatrix"</span>, 
  <span class="dt">reducedDimsOut =</span> <span class="st">"IterativeLSI"</span>,
  <span class="dt">threads =</span> threads
)

<span class="co">#Identify Clusters from Iterative LSI</span>
proj &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addClusters.html">addClusters</a></span>(<span class="dt">input =</span> proj, <span class="dt">reducedDims =</span> <span class="st">"IterativeLSI"</span>, <span class="dt">resolution =</span> <span class="fl">1.2</span>)

<span class="co">#Add Imputation Weights for Visualization</span>
proj &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addImputeWeights.html">addImputeWeights</a></span>(<span class="dt">ArchRProj =</span> proj)

<span class="co">#Compute a UMAP embedding to visualize our tiled matrix</span>
proj &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addEmbedding.html">addEmbedding</a></span>(
  <span class="dt">ArchRProj =</span> proj, 
  <span class="dt">reducedDims =</span> <span class="st">"IterativeLSI"</span>, 
  <span class="dt">embedding =</span> <span class="st">"UMAP"</span>, 
  <span class="dt">embeddingParams =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">min_dist =</span> <span class="fl">0.4</span>),
  <span class="dt">force =</span> <span class="ot">TRUE</span>
)

<span class="co">#Plot the UMAP Embedding with Metadata Overlayed</span>
plotList &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>()
plotList[[<span class="dv">1</span>]] &lt;-<span class="st"> </span><span class="kw"><a href="../reference/plotEmbedding.html">plotEmbedding</a></span>(<span class="dt">ArchRProj =</span> proj, <span class="dt">colorBy =</span> <span class="st">"colData"</span>, <span class="dt">name =</span> <span class="st">"Sample"</span>)
plotList[[<span class="dv">2</span>]] &lt;-<span class="st"> </span><span class="kw"><a href="../reference/plotEmbedding.html">plotEmbedding</a></span>(<span class="dt">ArchRProj =</span> proj, <span class="dt">colorBy =</span> <span class="st">"colData"</span>, <span class="dt">name =</span> <span class="st">"Clusters"</span>, <span class="dt">plotParams =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">labelMeans=</span><span class="ot">TRUE</span>))
<span class="kw"><a href="../reference/plotPDF.html">plotPDF</a></span>(<span class="dt">plotList =</span> plotList, <span class="dt">name =</span> <span class="st">"UMAP-Samples-Clusters"</span>, <span class="dt">width =</span> <span class="dv">6</span>, <span class="dt">height =</span> <span class="dv">6</span>, <span class="dt">ArchRProj =</span> proj)</code></pre></div>
<p>Using our tutorial data, your UMAP plots should look like <a href="articles/Articles/Tutorial_dimReduction_UMAP.pdf">this</a>.</p>
</div>
<div id="identifying-cluster-cell-types-using-marker-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#identifying-cluster-cell-types-using-marker-genes" class="anchor"></a>Identifying Cluster Cell Types Using Marker Genes</h3>
<p>In order to understand which clusters correspond to which cell types, we use a supervised approach based on prior knowledge of the genes that are active in specific cell types. We determine <em>gene activity scores</em> for each putative marker gene based on chromatin accessibility signal in the region surrounding the gene’s promoter. We can then overlay these <em>gene activity scores</em> on our UMAP embedding to visualize the relationship between gene activity and cluster. For more details, see the <a href="articles/Articles/geneScores.html">marker genes vignette</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">markerGenes  &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(
    <span class="st">"CD34"</span>, <span class="co">#Early Progenitor</span>
    <span class="st">"GATA1"</span>, <span class="co">#Erythroid</span>
    <span class="st">"PAX5"</span>, <span class="st">"MS4A1"</span>, <span class="co">#B-Cell Trajectory</span>
    <span class="st">"CD14"</span>, <span class="co">#Monocytes</span>
    <span class="st">"CD3D"</span>, <span class="st">"CD8A"</span>, <span class="st">"TBX21"</span>, <span class="st">"IL7R"</span> <span class="co">#TCells</span>
  )

<span class="co">#Plot the UMAP Embedding with Marker Genes Overlayed</span>
plotList &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>()
plotList[[<span class="dv">1</span>]] &lt;-<span class="st"> </span><span class="kw"><a href="../reference/plotEmbedding.html">plotEmbedding</a></span>(<span class="dt">ArchRProj =</span> proj, <span class="dt">colorBy =</span> <span class="st">"GeneScoreMatrix"</span>, <span class="dt">name =</span> markerGenes)
<span class="kw"><a href="../reference/plotPDF.html">plotPDF</a></span>(<span class="dt">plotList =</span> plotList, <span class="dt">name =</span> <span class="st">"UMAP-Marker-Gene-Scores"</span>, <span class="dt">width =</span> <span class="dv">6</span>, <span class="dt">height =</span> <span class="dv">6</span>, <span class="dt">ArchRProj =</span> proj)

<span class="co">#Plot Tracks at Marker Genes</span>
plotTracks &lt;-<span class="st"> </span><span class="kw"><a href="../reference/ArchRRegionTrack.html">ArchRRegionTrack</a></span>(<span class="dt">ArchRProj =</span> proj, <span class="dt">threads =</span> threads, <span class="dt">geneSymbol =</span> markerGenes)
<span class="kw"><a href="../reference/plotPDF.html">plotPDF</a></span>(<span class="dt">plotList =</span> plotTracks, <span class="dt">name =</span> <span class="st">"Tracks-Marker-Genes"</span>, <span class="dt">width =</span> <span class="dv">6</span>, <span class="dt">height =</span> <span class="dv">8</span>, <span class="dt">ArchRProj =</span> proj)

<span class="co">#Identify Marker Gene through Pairwise Test vs Null</span>
markersGS &lt;-<span class="st"> </span><span class="kw"><a href="../reference/markerFeatures.html">markerFeatures</a></span>(<span class="dt">ArchRProj =</span> proj, <span class="dt">threads =</span> threads, <span class="dt">useMatrix =</span> <span class="st">"GeneScoreMatrix"</span>)
heatmapGS &lt;-<span class="st"> </span><span class="kw"><a href="../reference/markerHeatmap.html">markerHeatmap</a></span>(
  <span class="dt">seMarker =</span> markersGS, 
  <span class="dt">cutOff =</span> <span class="st">"FDR &lt;= 0.01 &amp; Log2FC &gt;= 1"</span>, 
  <span class="dt">labelMarkers =</span> markerGenes
)
<span class="kw"><a href="../reference/plotPDF.html">plotPDF</a></span>(heatmapGS, <span class="dt">name =</span> <span class="st">"GS-Marker-Heatmap"</span>, <span class="dt">width =</span> <span class="dv">8</span>, <span class="dt">height =</span> <span class="dv">12</span>, <span class="dt">ArchRProj =</span> proj)</code></pre></div>
<p>Using our tutorial data, your gene activity score UMAP plots should look like <a href="articles/Articles/Tutorial_geneScores_UMAP.pdf">this</a>.</p>
</div>
<div id="creating-a-reproducible-peak-set" class="section level3">
<h3 class="hasAnchor">
<a href="#creating-a-reproducible-peak-set" class="anchor"></a>Creating a Reproducible Peak Set</h3>
<p>One of the most complicated aspects about ATAC-seq and scATAC-seq analysis is the generation of a reproducible and robust peak set. In ArchR, we use an iterative overlap removal process that we first described in <a href="https://www.ncbi.nlm.nih.gov/pubmed/30361341">Corces* &amp; Granja* et al. Science 2018</a>. This process is described in detail in the <a href="articles/Articles/peakCalling.html">peak calling vignette</a>.</p>
<p>To robustly call peaks, we first merge the sparse single-cell data into pseudobulks by aggregating the insertions from many individual cells into a single group. We make multiple pseudobulks for each cluster to enable an assessment of peak reproducibility. This process of pseudobulk generation is described in detail in the <a href="articles/Articles/pseudobulkGeneration.html">pseudobulk generation vignette</a>. We than call peaks using <a href="https://github.com/taoliu/MACS">MACS2 v???</a> and perform our iterative overlap removal. Once we obtain a finalized peak set, we obtain insertion counts in each peak for each single cell and associate this with the corresponding <code>ArrowFile</code> via the <code>ArchRProject</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Create Group Coverage Files that can be used for downstream analysis (~5-10 minutes)</span>
proj &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addGroupCoverages.html">addGroupCoverages</a></span>(<span class="dt">ArchRProj =</span> proj, <span class="dt">groupBy =</span> <span class="st">"Clusters"</span>, <span class="dt">threads =</span> threads)

<span class="co">#Call Reproducible Peaks w/ Macs2 (~5-10 minutes)</span>
proj &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addReproduciblePeakSet.html">addReproduciblePeakSet</a></span>(<span class="dt">ArchRProj =</span> proj, <span class="dt">groupBy =</span> <span class="st">"Clusters"</span>, <span class="dt">threads =</span> threads)

<span class="co">#Add Peak Matrix</span>
proj &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addPeakMatrix.html">addPeakMatrix</a></span>(<span class="dt">ArchRProj =</span> proj, <span class="dt">threads =</span> threads, <span class="dt">force =</span> <span class="ot">TRUE</span>)</code></pre></div>
</div>
<div id="identifying-marker-peaks" class="section level3">
<h3 class="hasAnchor">
<a href="#identifying-marker-peaks" class="anchor"></a>Identifying Marker Peaks</h3>
<p>Often times, we are interested to know which peaks are unique to an individual cluster or a small group of clusters. We can do this in an unsupervised fashion in ArchR:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Identify Marker Peaks</span>
markersPeaks &lt;-<span class="st"> </span><span class="kw"><a href="../reference/markerFeatures.html">markerFeatures</a></span>(<span class="dt">ArchRProj =</span> proj, <span class="dt">threads =</span> threads, <span class="dt">useMatrix =</span> <span class="st">"PeakMatrix"</span>)
heatmapPeaks &lt;-<span class="st"> </span><span class="kw"><a href="../reference/markerHeatmap.html">markerHeatmap</a></span>(
  <span class="dt">seMarker =</span> markersPeaks, 
  <span class="dt">cutOff =</span> <span class="st">"FDR &lt;= 0.01 &amp; Log2FC &gt;= 1"</span>
)
<span class="kw"><a href="../reference/plotPDF.html">plotPDF</a></span>(heatmapPeaks, <span class="dt">name =</span> <span class="st">"Peak-Marker-Heatmap"</span>, <span class="dt">width =</span> <span class="dv">8</span>, <span class="dt">height =</span> <span class="dv">12</span>, <span class="dt">ArchRProj =</span> proj)</code></pre></div>
<p>Using our tutorial data, your marker peak heatmap should look like <a href="articles/Articles/Tutorial_markerPeaks_HEATMAP.pdf">this</a>.</p>
</div>
<div id="performing-motif-enrichments" class="section level3">
<h3 class="hasAnchor">
<a href="#performing-motif-enrichments" class="anchor"></a>Performing Motif Enrichments</h3>
<p>Using the reproducible peak set that we defined above, we can use ArchR to calculate TF deviations on a single-cell basis for transcription factors in the peaks identified in each cluster. We can then overlay these deviations on on UMAP embedding. This effectively infers differences in TF activity across all single cells and is very useful in identifying regulatory factors governing cell fate.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Motif Search in PeakSet</span>
proj &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addMotifAnnotations.html">addMotifAnnotations</a></span>(<span class="dt">ArchRProj =</span> proj, <span class="dt">name =</span> <span class="st">"Motif"</span>)

<span class="co">#Add chromVAR Deviations (~20-25 min if using CisBP MotifSet)</span>
proj &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addDeviationsMatrix.html">addDeviationsMatrix</a></span>(<span class="dt">ArchRProj =</span> proj, <span class="dt">threads =</span> threads)

<span class="co">#Identify Variable TFs</span>
<span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw">getVarDeviations</span>(proj, <span class="dt">plot =</span> <span class="ot">FALSE</span>)$name, <span class="dv">25</span>)

<span class="co">#To access motif need to specify deviations,z : motif_name</span>
<span class="co">#Try getFeatures with MotifMatrix to see available names</span>
<span class="kw"><a href="../reference/getFeatures.html">getFeatures</a></span>(proj, <span class="dt">select =</span> <span class="st">"PAX5"</span>, <span class="dt">useMatrix =</span> <span class="st">"MotifMatrix"</span>)

<span class="co">#Define the list of motifs to plot</span>
motifs &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"GATA1_383"</span>, <span class="st">"CEBPA_155"</span>, <span class="st">"EBF1_67"</span>,<span class="st">"IRF4_632"</span>,<span class="st">"TBX21_780"</span>,<span class="st">"PAX5_709"</span>)

<span class="co">#Plot the UMAP Embedding with chromVAR Deviations Overlayed</span>
plotList &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>()
plotList[[<span class="dv">1</span>]] &lt;-<span class="st"> </span><span class="kw"><a href="../reference/plotEmbedding.html">plotEmbedding</a></span>(<span class="dt">ArchRProj =</span> proj, <span class="dt">colorBy =</span> <span class="st">"colData"</span>, <span class="dt">name =</span> <span class="st">"Clusters"</span>)
plotList[[<span class="dv">2</span>]] &lt;-<span class="st"> </span><span class="kw"><a href="../reference/plotEmbedding.html">plotEmbedding</a></span>(<span class="dt">ArchRProj =</span> proj, <span class="dt">colorBy =</span> <span class="st">"MotifMatrix"</span>, <span class="dt">name =</span> <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"z:"</span>,motifs))
<span class="kw"><a href="../reference/plotPDF.html">plotPDF</a></span>(<span class="dt">plotList =</span> plotList, <span class="dt">name =</span> <span class="st">"Plot-UMAP-TileLSI-MotifMatrix"</span>, <span class="dt">width =</span> <span class="dv">6</span>, <span class="dt">height =</span> <span class="dv">6</span>, <span class="dt">ArchRProj =</span> proj)</code></pre></div>
<p>Using our tutorial data, your marker peak heatmap should look like <a href="articles/Articles/Tutorial_motifEnrichment_UMAP.pdf">this</a>.</p>
</div>
<div id="performing-tf-footprinting" class="section level3">
<h3 class="hasAnchor">
<a href="#performing-tf-footprinting" class="anchor"></a>Performing TF Footprinting</h3>
<p>Transcription factor footprinting can also be done in ArchR with a single command. We note that the footprints generated by the tutorial data are not as clean as would be desired but this is because of the small size of the tutorial dataset.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Plot Motif Footprints from positions list</span>
<span class="co">#Recommend doing a few motifs not entire motif set</span>
seFoot &lt;-<span class="st"> </span><span class="kw"><a href="../reference/plotFootprints.html">plotFootprints</a></span>(proj, <span class="dt">positions =</span> <span class="kw"><a href="../reference/getPositions.html">getPositions</a></span>(proj)[motifs])</code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://twitter.com/JeffreyGranja">Jeffrey Granja</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
