<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Add a Reproducible Peak Set to an ArchRProject — addReproduciblePeakSet • ArchR</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />
<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Add a Reproducible Peak Set to an ArchRProject — addReproduciblePeakSet" />

<meta property="og:description" content="This function will get insertions from coverage files, call peaks,
and merge peaks to get a &quot;Union Reproducible Peak Set&quot;" />

<meta property="og:image" content="/logo.png" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ArchR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/Articles/tutorial.html">Tutorial</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Articles/doubletRemoval.html">Doublet Removal</a>
    </li>
    <li>
      <a href="../articles/Articles/qcFilters.html">QC Filters</a>
    </li>
    <li>
      <a href="../articles/Articles/peakCalling.html">Peak Calling</a>
    </li>
    <li>
      <a href="../articles/Articles/dimReduction.html">Dimensionality Reduction</a>
    </li>
    <li>
      <a href="../articles/Articles/geneScores.html">Gene Activity Scores</a>
    </li>
    <li>
      <a href="../articles/Articles/markerFeatures.html">Marker Feature Identification</a>
    </li>
    <li>
      <a href="../articles/Articles/signalTracks.html">Signal Track Ploting</a>
    </li>
    <li>
      <a href="../articles/Articles/footprinting.html">TF Footprinting</a>
    </li>
    <li>
      <a href="../articles/Articles/pseudobulkGeneration.html">Pseudobulk Replicate Generation</a>
    </li>
    <li>
      <a href="../articles/Articles/annotations.html">Gene and Genome Annotation</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/Articles/faq.html">FAQ</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/GreenleafLab/ArchR">
    <span class="fa fa-github"></span>
     
    ArchR on GitHub
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Add a Reproducible Peak Set to an ArchRProject</h1>
    
    <div class="hidden name"><code>addReproduciblePeakSet.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>This function will get insertions from coverage files, call peaks,
and merge peaks to get a "Union Reproducible Peak Set"</p>
    
    </div>

    <pre class="usage"><span class='fu'>addReproduciblePeakSet</span>(<span class='kw'>ArchRProj</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>groupBy</span> <span class='kw'>=</span> <span class='st'>"Clusters"</span>,
  <span class='kw'>reproducibility</span> <span class='kw'>=</span> <span class='st'>"2"</span>, <span class='kw'>peaksPerCell</span> <span class='kw'>=</span> <span class='fl'>500</span>, <span class='kw'>maxPeaks</span> <span class='kw'>=</span> <span class='fl'>150000</span>,
  <span class='kw'>minCells</span> <span class='kw'>=</span> <span class='fl'>25</span>, <span class='kw'>excludeChr</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"chrM"</span>, <span class='st'>"chrY"</span>),
  <span class='kw'>pathToMacs2</span> <span class='kw'>=</span> <span class='fu'><a href='findMacs2.html'>findMacs2</a></span>(), <span class='kw'>genomeSize</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>shift</span> <span class='kw'>=</span> -<span class='fl'>75</span>,
  <span class='kw'>extsize</span> <span class='kw'>=</span> <span class='fl'>150</span>, <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"q"</span>, <span class='kw'>cutOff</span> <span class='kw'>=</span> <span class='fl'>0.1</span>, <span class='kw'>extendSummits</span> <span class='kw'>=</span> <span class='fl'>250</span>,
  <span class='kw'>promoterDist</span> <span class='kw'>=</span> <span class='fl'>500</span>, <span class='kw'>genomeAnno</span> <span class='kw'>=</span> <span class='fu'><a href='getGenomeAnnotation.html'>getGenomeAnnotation</a></span>(<span class='no'>ArchRProj</span>),
  <span class='kw'>geneAnno</span> <span class='kw'>=</span> <span class='fu'><a href='getGeneAnnotation.html'>getGeneAnnotation</a></span>(<span class='no'>ArchRProj</span>),
  <span class='kw'>additionalParams</span> <span class='kw'>=</span> <span class='st'>"--nomodel --nolambda"</span>, <span class='kw'>threads</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>parallelParam</span> <span class='kw'>=</span> <span class='st'>"mclapply"</span>, <span class='kw'>force</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>verboseHeader</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>verboseAll</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>ArchRProj</th>
      <td><p>An <code>ArchRProject</code> object.</p></td>
    </tr>
    <tr>
      <th>groupBy</th>
      <td><p>The name of the column in <code>cellColData</code> to use for grouping cells together for peak calling.</p></td>
    </tr>
    <tr>
      <th>reproducibility</th>
      <td><p>A string that indicates how peak reproducibility should be handled. This string is dynamic and can be a function of <code>n</code> where <code>n</code> is the number of samples being assessed. For example, <code>reproducibility = "2"</code> means at least 2 samples must have a peak call at this locus and <code>reproducibility = "(n+1)/2"</code> means that the majority of samples must have a peak call at this locus.</p></td>
    </tr>
    <tr>
      <th>peaksPerCell</th>
      <td><p>The limit of number of peaks that can be identified per cell (this is useful for controlling how many peaks can be called from low cell groups).</p></td>
    </tr>
    <tr>
      <th>maxPeaks</th>
      <td><p>A numeric threshold for the maximum peaks to retain per group in <code>groupBy</code> in the union reproducible peak set.</p></td>
    </tr>
    <tr>
      <th>minCells</th>
      <td><p>The minimum number of unique cells that was used to create the coverage files on which peaks are called. This is important to allow for exclusion of pseudo-bulk replicates derived from very low cell numbers.</p></td>
    </tr>
    <tr>
      <th>excludeChr</th>
      <td><p>A character vector containing the <code>seqnames</code> of the chromosomes that should be excluded from peak calling.</p></td>
    </tr>
    <tr>
      <th>pathToMacs2</th>
      <td><p>The full path to the MACS2 executable.</p></td>
    </tr>
    <tr>
      <th>genomeSize</th>
      <td><p>The genome size to be used for MACS2 peak calling (see MACS2 documentation).</p></td>
    </tr>
    <tr>
      <th>shift</th>
      <td><p>The number of basepairs to shift each Tn5 insertion. When combined with <code>extsize</code> this allows you to create proper fragments, centered at the Tn5 insertion site, for use with MACS2 (see MACS2 documentation).</p></td>
    </tr>
    <tr>
      <th>extsize</th>
      <td><p>The number of basepairs to extend the MACS2 fragment after <code>shift</code> has been applied. When combined with <code>extsize</code> this allows you to create proper fragments, centered at the Tn5 insertion site, for use with MACS2 (see MACS2 documentation).</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>The method to use for significance testing in MACS2. Options are "p" for p-value and "q" for q-value. When combined with <code>cutOff</code> this gives the method and significance threshold for peak calling (see MACS2 documentation).</p></td>
    </tr>
    <tr>
      <th>cutOff</th>
      <td><p>The numeric significance cutoff for the testing method indicated by <code>method</code> (see MACS2 documentation).</p></td>
    </tr>
    <tr>
      <th>extendSummits</th>
      <td><p>The number of basepairs to extend peak summits (in both directions) to obtain final fixed-width peaks. For example, <code>extendSummits = 250</code> will create 501-bp fixed-width peaks from the 1-bp summits.</p></td>
    </tr>
    <tr>
      <th>promoterDist</th>
      <td><p>The maximum distance in basepairs from the peak summit to the nearest transcriptional start site to allow for a peak to be annotated as a "promoter" peak.</p></td>
    </tr>
    <tr>
      <th>genomeAnno</th>
      <td><p>The genomeAnnotation (see createGenomeAnnotation) is used for downstream analyses for genome information such as nucleotide information (GC info) or chromosome sizes.</p></td>
    </tr>
    <tr>
      <th>geneAnno</th>
      <td><p>The geneAnnotation (see createGeneAnnotation) is used for peak labeling as promoter etc.</p></td>
    </tr>
    <tr>
      <th>additionalParams</th>
      <td><p>A string of additional parameters to pass to MACS2 (see MACS2 documentation).</p></td>
    </tr>
    <tr>
      <th>threads</th>
      <td><p>The number of threads to be used for parallel computing.</p></td>
    </tr>
    <tr>
      <th>parallelParam</th>
      <td><p>A list of parameters to be passed for biocparallel/batchtools parallel computing.</p></td>
    </tr>
    <tr>
      <th>force</th>
      <td><p>A boolean value indicating whether to force the reproducible peak set to be overwritten if it already exist in the given <code>ArchRProject</code> peakSet.</p></td>
    </tr>
    <tr>
      <th>verboseHeader</th>
      <td><p>A boolean value that determines whether standard output includes verbose sections.</p></td>
    </tr>
    <tr>
      <th>verboseAll</th>
      <td><p>A boolean value that determines whether standard output includes verbose subsections.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>additional args</p></td>
    </tr>
    </table>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
                </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://twitter.com/JeffreyGranja'>Jeffrey Granja</a>, <a href='https://twitter.com/doctorcorces'>Ryan Corces</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


